EZP-27458: Aggregation API (#1171)
ezsystems/developer-documentation: docs/guide/search/extend_elasticsearch.md

                - { name: ezplatform.search.elasticsearch.query.location.sort_clause_visitor }
    ```

    ## Custom Aggregation
    
    To create a custom aggregation for use with Elasticsearch, create an aggregation class.

>     In the following example it is an aggregation that groups Location query results according to the Location priority:

Reviewer: ```suggestion
    In the following example, an aggregation groups Location query results according to the Location priority:
```

    
    ``` php
    <?php

    declare(strict_types=1);

    namespace App\Query\Aggregation;

    use eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractRangeAggregation;
    use eZ\Publish\API\Repository\Values\Content\Query\Aggregation\LocationAggregation;

    final class PriorityRangeAggregation extends AbstractRangeAggregation implements LocationAggregation
    {

    }

>     ```

Reviewer 2: The most important information is missing here: what is and what does the aggregation class. 

For instance:

> To provide support for custom aggregation, you will need to create aggregation definition by implementing  `\eZ\Publish\API\Repository\Values\Content\Query\Aggregation` interface  or inherit one of following abstract classes:
> 
> *   `\eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractRangeAggregation`
> *   `\eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractStatsAggregation`
> *   `\eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractTermAggregation`
> 
> Aggregation definition needs contain at least aggregation name and optional aggregation parameters e.g. path string used to limit aggregation results to specific subtree or content type identifier / field definition identifier which will be mapped latter on to search index field name. Aggregation definition should be independent of used search engine....

Latter on we should describe propose of 

*   `\eZ\Publish\API\Repository\Values\Content\Query\Aggregation\FieldAggregation`
*   `\eZ\Publish\API\Repository\Values\Content\Query\Aggregation\LocationAggregation`
*   `\eZ\Publish\API\Repository\Values\Content\Query\Aggregation\RawAggregation`

interfaces, explain used naming convention and then we can go with example :wink:

>     A custom aggregation requires an aggregation visitor and a result extractor.

Reviewer 2: Here we also need explanation of what is aggregation visitor / result extractor as well as the explanation of its role and responsibilities.

    
    For simpler cases, you can use one of the built-in visitors corresponding to the aggregation type.
    In the example below it is `RangeAggregationVisitor`:
    
    ``` yaml
    services:
        App\Query\Aggregation\Elasticsearch\PriorityAggregationVisitor:
            class: Ibexa\Platform\ElasticSearchEngine\Query\AggregationVisitor\RangeAggregationVisitor
            factory: ['@Ibexa\Platform\ElasticSearchEngine\Query\AggregationVisitor\Factory\SearchFieldAggregationVisitorFactory', 'createRangeAggregationVisitor']
            arguments:
                $aggregationClass: 'App\Query\Aggregation\PriorityRangeAggregation'
                $searchIndexFieldName: 'priority_i'
            tags:
                - { name: ezplatform.search.elasticsearch.query.location.aggregation_visitor }
    ```
    
    The visitor is created by `SearchFieldAggregationVisitorFactory`.
    You provide it with two arguments:
    
    - the aggregation class in `aggregationClass`
    - the field name in search index in `searchIndexFieldName`
    
    Tag the service with `ezplatform.search.elasticsearch.query.location.aggregation_visitor`.
    

>     For the result extractor you can use the built-in `RangeAggregationResultExtractor`

Reviewer: ```suggestion
    For the result extractor, you can use the built-in `RangeAggregationResultExtractor`
```

    use eZ\Publish\API\Repository\Values\Content\Query\Aggregation;
    use eZ\Publish\API\Repository\Values\Content\Search\AggregationResult;
    use Ibexa\Platform\Contracts\ElasticSearchEngine\Query\AggregationResultExtractor;
    use Ibexa\Platform\Contracts\ElasticSearchEngine\Query\LanguageFilter;
    use stdClass;

    final class PriorityAggregationResultExtractor implements AggregationResultExtractor
    {

        public function supports(Aggregation $aggregation, LanguageFilter $languageFilter): bool
        {
            return $aggregation instanceof PriorityRangeAggregation;
        }

        public function extract(Aggregation $aggregation, LanguageFilter $languageFilter, array $data): AggregationResult
        {
            $entries = [];

            foreach ($data['buckets'] as $bucket) {
                $entries[] = new AggregationResult\RangeAggregationResultEntry(
                    new Aggregation\Range($bucket['from'] ?? null, $bucket['to'] ?? null),
                    $bucket['doc_count']
                );
            }

            return new AggregationResult\RangeAggregationResult($aggregation->getName(), $entries);
        }
    }
    ```
    

>     The `supports()` method check whether the provided aggregation is of the supported type

Reviewer: ```suggestion
    The `supports()` method checks whether the provided aggregation is of the supported type
```

    (in this case, your custom `PriorityRangeAggregation`).
    

>     The `extract()` method converts the [JSON-encoded data provided by the search engine](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations.html) to a `RangeAggregationResult` object.

Reviewer 2: ```suggestion
    The `extract()` method converts the [raw data provided by the search engine](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations.html) to a `RangeAggregationResult` object.
```

At this stage data is already decoded from JSON :wink:

    
    Finally, register both the aggregation visitor and the result extractor as services.
    
    Tag the aggregation visitor with `ezplatform.search.elasticsearch.query.location.aggregation_visitor`

>     and the result extractor with `ezplatform.search.elasticsearch.query.location.aggregation_result_extractor`:

Reviewer 2: Tags 

*   `ezplatform.search.elasticsearch.query.content.aggregation_visitor`
*   `ezplatform.search.elasticsearch.query.content.aggregation_result_extractor`

needs to mentioned and described in documentation as well.

    In the following example, an aggregation groups Location query results according to the Location priority:
    
    ``` php
    <?php

    declare(strict_types=1);

    namespace App\Query\Aggregation;

    use eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractRangeAggregation;
    use eZ\Publish\API\Repository\Values\Content\Query\Aggregation\LocationAggregation;

    final class PriorityRangeAggregation extends AbstractRangeAggregation implements LocationAggregation
    {

    }
    ```

    PriorityRangeAggregation` in the example above extends `AbstractRangeAggregation`.
    The name indicates that it is going to aggregate the results according to the Location priority, using Range aggregation.
    
    An aggregation must implement the `eZ\Publish\API\Repository\Values\Content\Query\Aggregation` interface or inherit one of following abstract classes:
    
    - `eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractRangeAggregation`
    - `eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractStatsAggregation`
    - `eZ\Publish\API\Repository\Values\Content\Query\Aggregation\AbstractTermAggregation`
    

>     An aggregation should also implement one of the following interfaces:

Reviewer 2: ```suggestion
    An aggregation can also implement one of the following interfaces:
```
